SQL: Allow text formats to work in async mode

bpintea opened the pull request:
This extends the async implementation to support working with the text
formats (txt, csv, tsv).

A test validating the administrator operation of a user with the
"manage" permission has also been added.

bpintea committed:
Allow text formats to work in async mode

This extends the async implementation to support working with the text
formats (txt, csv, tsv).

A test validating the administrator operation of a user with the
"manage" permission has also been added.

elasticmachine commented:
Pinging @elastic/es-ql (Team:QL)

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/RestSqlAsyncGetResultsAction.java:
```diff
@@ -40,6 +42,12 @@ protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient cli
+
+    @Override
+    protected Set<String> responseParams() {
+        return Collections.singleton(URL_PARAM_DELIMITER);
```
This seems strange, why do we return the Delimiter?

bpintea replied:
That's used to signal that this param can be checked on, but not consumed. [BaseRestHandler.java#prepareRequest](https://github.com/elastic/elasticsearch/blob/c6c2f1bc3637433db7791414983f9cd1911399b0/server/src/main/java/org/elasticsearch/rest/BaseRestHandler.java#L159) and responseParams() details how that works.

matriv commented on x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/rest/RestSqlTestCase.java:
```diff
@@ -1137,43 +1148,54 @@ private void executeQueryWithNextPage(String format, String expectedHeader, Stri
+        }
+        request.setJsonEntity(bulk.toString());
+        client().performRequest(request);
+
```
nit: extra empty line.

bpintea replied:
Thx, will remove.

bpintea replied:
Removed.

matriv commented on x-pack/plugin/sql/qa/server/security/src/test/java/org/elasticsearch/xpack/sql/qa/security/RestSqlSecurityAsyncIT.java:
```diff
@@ -101,6 +102,24 @@ private void testCase(String user, String otherUser) throws Exception {
+
+        Response deleteResp = deleteAsyncSqlSearch(id, "manage_user");
+        assertOK(deleteResp);
+
```
nit: extra empty line

bpintea replied:
Removed.

matriv commented on x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/rest/RestSqlTestCase.java:
```diff
@@ -1183,4 +1205,207 @@ public static void assertResponse(Map<String, Object> expected, Map<String, Obje
+        expected.put(IS_RUNNING_NAME, false);
+        expected.put(COLUMNS_NAME, singletonList(columnInfo(mode, "1", "integer", JDBCType.INTEGER, 11)));
+        expected.put(ROWS_NAME, singletonList(singletonList(1)));
+        assertAsyncResponse(expected, runSql(builder, mode));
```
I'm worrying again here and for the rest of the aysnc tests about flakiness. Should we use the `assertBusy` mechanism to make sure that it always works?

bpintea replied:
The `assertBusy` could be used in this case, true (tho with extra complexity). How would you see it failing in a flaky way?

matriv replied:
Not easy to achieve I guess, If you want to just check locally maybe try a simple `Thread.sleep()` somewhere in the SQL chain (e.g. Analyzer), so that the execution is delayed for let's say 5secs.

bpintea replied:
IIUC your suggestion, the (merged) local [cancellation tests](https://github.com/elastic/elasticsearch/blob/feat/sql-async/x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/CancellationTests.java) do that, they start a task and make sure it blocks at certain stages along its execution.

The tests here are all remote ones though, that are only mean to check the correct response formatting of the textual answers. The timing is irrelevant, it makes no difference between "10s" or "1d", for as long as the query finishes. But even if it doesn't finish, or the query goes wrong, it'd result in a failed test due to mismatch between expected and returned answer. That's why I wasn't sure how you see it failing in a flaky way.

Using `assertBusy` to check on task's actual completion before checking its result is possible, but don't think it's necessary. But let me know if you see it differently still.

matriv replied:
@bpintea Thx Bogdan, I was confused with the functionality here.
Since we have the `waitForCompletionTimeout` that ensures that we have the response (or a timeout if the test runs for 1 day :D) when we try to assert it, so no need for the `assertBusy`.

matriv reviewed (commented):
Thanks Bogdan, left some comments, mostly worrying about flakiness because of the async nature of things.

matriv commented on x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/rest/RestSqlTestCase.java:
```diff
@@ -1183,4 +1205,207 @@ public static void assertResponse(Map<String, Object> expected, Map<String, Obje
+        return runSqlAsTextFormat(builder, null);
+    }
+
+    static Response runSqlAsTextFormat(RequestObjectBuilder builder, @Nullable String format) throws IOException {
```
maybe a better name is `runSqlWithFormat`

bpintea replied:
Makes sense, will change.

bpintea replied:
Changed. to `runSqlAsTextWithFormat` since there are existing `runSqlAsText` methods and only text formats need to be provided (as it checks for a `Cursor` header, only available with those formats).

bpintea commented on x-pack/plugin/sql/qa/server/security/src/test/java/org/elasticsearch/xpack/sql/qa/security/RestSqlSecurityAsyncIT.java:
```diff
@@ -101,6 +102,24 @@ private void testCase(String user, String otherUser) throws Exception {
 
+    // user with manage privilege can check status and delete
+    public void testWithManager() throws IOException {
+        Response submitResp = submitAsyncSqlSearch("SELECT event_type FROM \"index\" WHERE val=0", TimeValue.timeValueSeconds(10), "user1");
```
This particular value has been c&p'ed from the similar [EQL tests](https://github.com/elastic/elasticsearch/blob/c6c2f1bc3637433db7791414983f9cd1911399b0/x-pack/plugin/eql/qa/security/src/javaRestTest/java/org/elasticsearch/xpack/eql/AsyncEqlSecurityIT.java#L66).

But the way the test works is to get exceptions when the "other" user tries to get to this async request. For that to work, the async task would only need to be created. Which is checked both by asserting an OK answer to the initial request, as well as by getting its status, subsequently. So the task wouldn't even have to be finished, just started, which is guaranteed by the first checks.

So the "10s" aren't actually relevant for the outcome of the test.

matriv replied:
Great, thx!

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/SqlResponseFormatter.java:
```diff
@@ -0,0 +1,95 @@
+
+class SqlResponseFormatter extends RestResponseListener<SqlQueryResponse> {
+
+    private final long startNanos = System.nanoTime();
```
That's how we counted [before](https://github.com/elastic/elasticsearch/blob/c6c2f1bc3637433db7791414983f9cd1911399b0/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/RestSqlQueryAction.java#L80) -- it's pretty much when the request enters SQL, but yeh, otherwise maybe "a bit" later than when it enters ES.

matriv replied:
ok, thx!

costin commented on x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/RestSqlIT.java:
```diff
@@ -85,7 +85,7 @@ public void testIncorrectAcceptHeader() throws IOException {
         expectBadRequest(
             () -> toMap(client().performRequest(request), "plain"),
-            containsString("Invalid response content type: Accept=[application/fff]")
+            containsString("Invalid request content type: Accept=[application/fff]")
```
👍

costin commented on x-pack/plugin/sql/sql-action/src/main/java/org/elasticsearch/xpack/sql/action/SqlQueryResponse.java:
```diff
@@ -194,7 +194,7 @@ public void writeTo(StreamOutput out) throws IOException {
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject();
         {
-            if (Strings.hasText(asyncExecutionId)) {
```
I wonder if using different class hierarchy instead of dedicated properties might help encapsulate the difference in response between sync/async better.
The `hasId` is obscure - is it used anywhere else?

bpintea replied:
> I wonder if using different class hierarchy instead of dedicated properties might help encapsulate the difference in response between sync/async better.

That'd be ideal, but I didn't find a way to cross the hierarchies.

> The hasId is obscure - is it used anywhere else?

`hasId()` is also used in SqlResponseFormatter and TextFormat, yes. It is a bit obscure and I hesitated using it, but eventually did since there is a `hasCursor()` (which returns the `cursor` field) and there is already the `id()` method. Now the `id()` method is maybe also generic, but I wanted to keep the symmetry with `EqlSearchResponse`.
(We could change them in both classes to something like `asyncId()` & co., but I'd do that subsequently, if needed.)

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/SqlMediaTypeParser.java:
```diff
@@ -34,27 +36,47 @@
+        return getResponseMediaType(request);
+    }
+
+    public static MediaType getResponseMediaType(RestRequest request) {
```
This method should delegate to the one above or vice-versa. 
The `URL_PARAM_FORMAT` & co is shared across the two.

bpintea replied:
The check on `URL_PARAM_FORMAT` needs to stay in both functions, since the functions are called from different context and need to act differently on its presence. But I've refactored the functions to avoid the double check.

costin reviewed (approved):
LGTM. I left a comment on two methods that seem to share code.
The sync/async code paths complicate things though I don't have any good advice on how to move forward without splitting the code apart.
We just have to be careful and use proper formatting to delimitate the two and where possible use different methods.

matriv reviewed (approved):
LGTM, thx!

bpintea committed:
Address review comments

- refactor media type calculation to avoid redundant check on URL param.
- rename methods in tests.
- remove empty lines.

bpintea commented:
@elasticmachine run elasticsearch-ci/rest-compatibility

astefan commented on x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/rest/RestSqlTestCase.java:
```diff
@@ -1183,4 +1204,207 @@ public static void assertResponse(Map<String, Object> expected, Map<String, Obje
+                    long millis = 1;
+                    for (boolean isRunning = true; isRunning; Thread.sleep(millis *= 2)) {
+                        asyncStatus = toMap(client().performRequest(request), null);
+                        isRunning = (boolean) asyncStatus.get(IS_RUNNING_NAME);
```
Is there a danger here of this `for` not existing and running into a test suite timeout?

bpintea replied:
That would require the server so somehow hang with the search task alive (which I guess would also trigger a timeout).

astefan commented on x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/rest/RestSqlTestCase.java:
```diff
@@ -1183,4 +1204,207 @@ public static void assertResponse(Map<String, Object> expected, Map<String, Obje
+                if (randomBoolean()) {
+                    request.addParameter(URL_PARAM_FORMAT, format);
+                    if (format.equals("csv")) {
+                        csvDelimiter = ';';
```
If I understand this well, you use the default delimiter (`,`) but depending on a `randomBoolean` you change that (and is set explicitly) to `;`. I would move `Character csvDelimiter = ',';` closer to this `randomBoolean`. If I wouldn't have put the code in the IDE and look at what happens with the `csvDelimiter` it would have been more difficult to see what the logic is.

bpintea replied:
The `randomBoolean()` splits between passing the `format` as a URL parameter _or_ HTTP header. Then, if the chosen branch is to use the parameter _and_ the format - also randomly chosen - is `csv` then the splitting character is changed.
The `randomBoolean()` isn't solely responsible for the delimiter char, so not sure how to make it clearer, but if you have a suggestion I'll apply it.

astefan commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/RestSqlQueryAction.java:
```diff
@@ -60,54 +48,9 @@ protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient cli
             sqlRequest = SqlQueryRequest.fromXContent(parser);
         }
 
-        MediaType responseMediaType = sqlMediaTypeParser.getResponseMediaType(request, sqlRequest);
-        if (responseMediaType == null) {
-            String msg = String.format(Locale.ROOT, "Invalid response content type: Accept=[%s], Content-Type=[%s], format=[%s]",
-                request.header("Accept"), request.header("Content-Type"), request.param("format"));
-            throw new IllegalArgumentException(msg);
-        }
-
-        /*
-         * Special handling for the "delimiter" parameter which should only be
-         * checked for being present or not in the case of CSV format. We cannot
-         * override {@link BaseRestHandler#responseParams()} because this
-         * parameter should only be checked for CSV, not always.
-         */
-        if ((responseMediaType instanceof XContentType || ((TextFormat) responseMediaType) != TextFormat.CSV)
-            && request.hasParam(URL_PARAM_DELIMITER)) {
-            throw new IllegalArgumentException(unrecognized(request, Collections.singleton(URL_PARAM_DELIMITER), emptySet(), "parameter"));
-        }
-
-        long startNanos = System.nanoTime();
```
I am not sure I agree with this code being "externalized" in a separate class that is a `RestResponseListener` called `SqlResponseFormatter`. If it's a Listener, this should be reflected in its name imo. Also, why its a separate class? Maybe an internal class in RestSqlQueryAction if you really like to refactor the code (to me it looks like a forced refactoring without any real benefit, maybe it's just me)?

bpintea replied:
> [...] class that is a RestResponseListener called SqlResponseFormatter

Good point, I've renamed it to `SqlResponseListener`.

> why its a separate class?

The listener is used by both `RestSqlQueryAction` and `RestSqlAsyncGetResultsAction`.

astefan reviewed (commented):
Looks good in general. Left few comments.

bpintea committed:
Address review comments

- renamed listener;
- minor test refactorings.

bpintea commented on x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/rest/RestSqlTestCase.java:
```diff
@@ -1183,4 +1204,207 @@ public static void assertResponse(Map<String, Object> expected, Map<String, Obje
+            Character csvDelimiter = ',';
+
+            assertEquals(200, response.getStatusLine().getStatusCode());
+            assertEquals(response.getHeader(HEADER_NAME_ASYNC_PARTIAL), response.getHeader(HEADER_NAME_ASYNC_RUNNING));
```
> assertEquals is about comparing a (actual) value to something that's expected. 

Not sure, I might have taken editor's suggestion to "simplify" the code. The documentation is not "contractual" enough, IMO, but parameter naming might support your case. :-) So I see your point and I've refactored to `assertTrue()` (tho maybe a bit less easy to read).

> Just few lines above you used the right way for these assertions

This case differs from above since both "true" or "false" are valid, as long as both headers have the same value.

astefan reviewed (approved):
LGTM. Thank you for addressing my review.

bpintea committed:
Fix imports past master merge

- Fix Nullable and TimeValue imports after classes relocation.
